lib/snippets/kubernetes-deploy

```diff
@@ -75,90 +203,375 @@ class KubernetesDeploy
+      @found = st.success?
+      log_status(status_data)
+    end
+
+    def succeeded?
+      exists?
+    end
+
+    def failed?
+      false
+    end
+
+    def exists?
+      @found
+    end
+  end
+
+  class Service < KubernetesResource
+    def initialize(name, namespace, file)
+      @name, @namespace, @file = name, namespace, file
+    end
+
+    def sync
+      endpoints, st = run_kubectl("get", "endpoints", @name, "--output=jsonpath={.subsets[*].addresses[*].ip}")
+      @num_endpoints = endpoints.split(/\s+/).length
+      @status = "#{@num_endpoints} endpoints"
+      @found = st.success?
+      log_status(status_data)
+    end
+
+    def succeeded?
+      @num_endpoints > 0
+    end
+
+    def failed?
+      false
+    end
+
+    def exists?
+      @found
+    end
+  end
+
+  class Pod < KubernetesResource
+    def initialize(name, namespace, file, parent: nil)
+      @name, @namespace, @file, @parent = name, namespace, file, parent
+      @bare = !@parent
+    end
+
+    def sync
+      out, st = run_kubectl("get", type, @name, "-a", "--output=jsonpath={.spec.containers[*].name}{\"//\"}{.status.phase}")
+      @found = st.success?
+      if @found
+        container_list, @status = out.split("//")
+        @containers = container_list.split(/\s+/)
+      end
+      fetch_logs
+      log_status(status_data)
+    end
+
+    def succeeded?
+      @bare ? (@status == "Succeeded") : (@status == "Running")
+    end
+
+    def failed?
+      @status == "Failed"
+    end
+
+    def exists?
+      @found
+    end
+
+    def status_data
+      group = @bare ? "Bare pods" : @parent
+      super.merge(
+        group: group,
+        logs: @logs
+      )
+    end
+
+    private
+
+    def fetch_logs
+      @logs = {}
+      return if !@found || @containers.empty?
+
+      @containers.each do |container_name|
+        out, _ = run_kubectl("logs", @name, "--timestamps=true", "--since=#{STEP_COMPLETION_TIMEOUT_MINUTES}m")
+        next if out.empty?
+        @logs[container_name] = out
+        if @bare && deployed? && !@already_displayed
+          KubernetesDeploy.logger.info "Logs from #{id} container #{container_name}:"
+          STDOUT.puts "#{out}"
+          @already_displayed = true
+        end
+      end
+    end
+  end
+
+  class Deployment < KubernetesResource
+    def initialize(name, namespace, file)
+      @name, @namespace, @file = name, namespace, file
+    end
+
+    def sync
+      deployment_stats, st = run_kubectl("get", type, @name)
+      @found = st.success?
+      if @found
+        # deployment_stats looks like this:
+        # NAME                  DESIRED   CURRENT   UP-TO-DATE   AVAILABLE   AGE
+        # jobs                  1         1         1            1           43d
+        labels, statuses = deployment_stats.split("\n")
+        labels = labels.split(/\s+/)[1..4]
+        statuses = statuses.split(/\s+/)[1..4]
+        @rollout_data = labels.zip(statuses).to_h # { "DESIRED" => 1 } etc.
```

Comment:
Good question... it was because this was exactly what I wanted, and I was having trouble digging it out of the `get deployment` data. Looking again, it shouldn't be _too_ bad. I'll give it another shot.